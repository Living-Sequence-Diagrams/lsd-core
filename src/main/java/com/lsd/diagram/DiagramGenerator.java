package com.lsd.diagram;

import com.lsd.properties.LsdProperties;
import com.lsd.report.model.Participant;
import com.lsd.report.model.SequenceDiagram;
import com.lsd.events.SequenceEvent;
import com.mitchellbosecke.pebble.PebbleEngine;
import com.mitchellbosecke.pebble.template.PebbleTemplate;
import lombok.Builder;
import lombok.SneakyThrows;

import java.io.StringWriter;
import java.io.Writer;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static com.lsd.properties.LsdProperties.DIAGRAM_THEME;
import static java.util.stream.Collectors.toList;

@Builder
public class DiagramGenerator {
    private final PebbleEngine engine = new PebbleEngine.Builder().autoEscaping(false).build();
    private final PebbleTemplate compiledTemplate = engine.getTemplate("templates/sequence-uml.peb");

    private final Set<String> includes;
    private final List<Participant> participants;
    private final List<SequenceEvent> events;

    public SequenceDiagram sequenceDiagram() {
        String uml = generateSequenceUml();
        String svg = generateSequenceSvg(uml);
        return SequenceDiagram.builder()
                .uml(uml)
                .svg(svg)
                .build();
    }

    private String generateSequenceSvg(String markup) {
        SvgConverter svgConverter = new SvgConverter();

        return svgConverter.convert(markup);
    }

    @SneakyThrows
    private String generateSequenceUml() {
        Writer writer = new StringWriter();
        compiledTemplate.evaluate(writer, Map.of(
                "theme", LsdProperties.get(DIAGRAM_THEME),
                "includes", includes,
                "participants", participants,
                "events", events.stream()
                        .map(SequenceEvent::toMarkup)
                        .collect(toList())
        ));
        return writer.toString();
    }
}
