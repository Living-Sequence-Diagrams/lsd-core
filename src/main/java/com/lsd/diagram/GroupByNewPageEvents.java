package com.lsd.diagram;

import com.lsd.events.Newpage;
import com.lsd.events.PageTitle;
import com.lsd.events.SequenceEvent;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.function.BiConsumer;
import java.util.function.BinaryOperator;
import java.util.function.Function;
import java.util.function.Supplier;
import java.util.stream.Collector;

/**
 * Splits a given List of sequence events into sublists if NewPage events are found. The NewPage event is converted to
 * a PageTitle event (preserving the title).
 */
class GroupByNewPageEvents implements Collector<SequenceEvent, List<List<SequenceEvent>>, List<List<SequenceEvent>>> {
    @Override
    public Supplier<List<List<SequenceEvent>>> supplier() {
        return () -> {
            var lists = new ArrayList<List<SequenceEvent>>();
            lists.add(new ArrayList<>());
            return lists;
        };
    }

    @Override
    public BiConsumer<List<List<SequenceEvent>>, SequenceEvent> accumulator() {
        return (lists, sequenceEvent) -> {
            if (sequenceEvent instanceof Newpage) {
                var newGrouping = new ArrayList<SequenceEvent>();
                newGrouping.add(PageTitle.titled(((Newpage) sequenceEvent).getTitle()));
                lists.add(newGrouping);
            } else {
                var currentGrouping = lists.get(lists.size() - 1);
                currentGrouping.add(sequenceEvent);
            }
        };
    }

    @Override
    public BinaryOperator<List<List<SequenceEvent>>> combiner() {
        return null;
    }

    @Override
    public Function<List<List<SequenceEvent>>, List<List<SequenceEvent>>> finisher() {
        return lists -> lists;
    }

    @Override
    public Set<Characteristics> characteristics() {
        return Set.of();
    }
}
