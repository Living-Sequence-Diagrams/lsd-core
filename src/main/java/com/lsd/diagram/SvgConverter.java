package com.lsd.diagram;

import net.sourceforge.plantuml.FileFormatOption;
import net.sourceforge.plantuml.SourceStringReader;
import org.jdom.Document;
import org.jdom.JDOMException;
import org.jdom.input.SAXBuilder;
import org.jdom.output.Format;
import org.jdom.output.XMLOutputter;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.StringReader;
import java.io.StringWriter;

import static net.sourceforge.plantuml.FileFormat.SVG;

public class SvgConverter {

    public String convert(String markup) {
        return prettyPrint(createSvg(markup));
    }

    private String createSvg(String plantUmlMarkup) {
        SourceStringReader reader = new SourceStringReader(plantUmlMarkup);
        try (final ByteArrayOutputStream os = new ByteArrayOutputStream()) {
            reader.outputImage(os, new FileFormatOption(SVG, false));
            return os.toString();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private String prettyPrint(String xml) {
        try {
            SAXBuilder sb = new SAXBuilder();
            Document doc = sb.build(new StringReader(xml));
            StringWriter stringWriter = new StringWriter();
            new XMLOutputter(Format.getPrettyFormat()).output(doc, stringWriter);
            return stringWriter.toString();
        } catch (JDOMException | IOException e) {
            throw new RuntimeException(e);
        }
    }
}