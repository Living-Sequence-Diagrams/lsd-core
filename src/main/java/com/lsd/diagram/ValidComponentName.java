package com.lsd.diagram;

import org.apache.commons.text.WordUtils;

import static java.util.Arrays.stream;
import static java.util.stream.Collectors.joining;

public class ValidComponentName {
    public static final String ILLEGAL_CHARS = "[()/]";

    public static String of(String name) {
        return stream(name
                .replaceAll(ILLEGAL_CHARS, " ")
                .split(" ")
        ).map(WordUtils::capitalize)
                .collect(joining())
                .replaceAll(" ", "");
    }
}
