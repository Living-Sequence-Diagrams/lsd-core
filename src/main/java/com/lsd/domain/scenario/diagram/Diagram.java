package com.lsd.domain.scenario.diagram;

public interface Diagram {

    /**
     * @return Returns an SVG representation of the diagram
     */
    String toSvg();
}
