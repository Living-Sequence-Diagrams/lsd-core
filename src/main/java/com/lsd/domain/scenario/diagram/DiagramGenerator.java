package com.lsd.domain.scenario.diagram;

import com.lsd.domain.Participant;
import com.lsd.domain.scenario.events.Event;
import com.lsd.domain.scenario.events.Interaction;
import com.lsd.properties.LsdProperties;
import com.mitchellbosecke.pebble.PebbleEngine;
import com.mitchellbosecke.pebble.template.PebbleTemplate;
import lombok.Builder;
import lombok.SneakyThrows;

import java.io.StringWriter;
import java.io.Writer;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static com.lsd.properties.LsdProperties.DIAGRAM_THEME;
import static java.util.stream.Collectors.toList;

@Builder
public class DiagramGenerator {
    private final PebbleEngine engine = new PebbleEngine.Builder().autoEscaping(false).build();
    private final PebbleTemplate compiledTemplate = engine.getTemplate("templates/sequence-uml.peb");

    private final Set<String> includes;
    private final List<Participant> participants;
    private final List<Event> events;

    public SequenceDiagram sequenceDiagram() {
        return SequenceDiagram.builder()
                .uml(generateSequenceUml())
                .interactions(events.stream()
                        .filter(Interaction.class::isInstance)
                        .map(Interaction.class::cast)
                        .collect(toList()))
                .build();
    }

    @SneakyThrows
    private String generateSequenceUml() {
        Writer writer = new StringWriter();
        compiledTemplate.evaluate(writer, Map.of(
                "theme", LsdProperties.get(DIAGRAM_THEME),
                "includes", includes,
                "participants", participants,
                "events", events.stream()
                        .map(Event::toMarkup)
                        .collect(toList())
        ));
        return writer.toString();
    }
}
