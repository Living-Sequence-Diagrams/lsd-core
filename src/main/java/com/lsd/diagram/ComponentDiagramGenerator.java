package com.lsd.diagram;

import com.lsd.IdGenerator;
import com.lsd.events.Message;
import com.lsd.events.SequenceEvent;
import com.lsd.events.SynchronousResponse;
import com.lsd.properties.LsdProperties;
import com.lsd.report.model.Diagram;
import com.lsd.report.model.Participant;
import com.mitchellbosecke.pebble.PebbleEngine;
import com.mitchellbosecke.pebble.template.PebbleTemplate;
import lombok.Builder;
import lombok.SneakyThrows;

import java.io.StringWriter;
import java.io.Writer;
import java.util.List;
import java.util.Map;

import static com.lsd.properties.LsdProperties.DIAGRAM_THEME;
import static java.util.function.Predicate.not;
import static java.util.stream.Collectors.toList;

@Builder
public class ComponentDiagramGenerator {
    private final PebbleEngine engine = new PebbleEngine.Builder().autoEscaping(false).build();
    private final PebbleTemplate compiledTemplate = engine.getTemplate("templates/component-uml.peb");
    private final SvgConverter svgConverter = new SvgConverter();

    private final List<Participant> participants;
    private final List<SequenceEvent> events;
    private final IdGenerator idGenerator;

    public Diagram diagram() {
        String uml = generateUml();
        return Diagram.builder()
                .id(idGenerator.next())
                .uml(uml)
                .svg(generateSvg(uml))
                .build();
    }

    private String generateSvg(String markup) {
        return svgConverter.convert(markup);
    }

    @SneakyThrows
    private String generateUml() {
        Writer writer = new StringWriter();
        compiledTemplate.evaluate(writer, Map.of(
                "theme", LsdProperties.get(DIAGRAM_THEME),
                "participants", participants.stream()
                        .map(Participant::getMarkup)
                        .map(markup -> markup.replaceAll("^participant ", "component "))
                        .distinct()
                        .collect(toList()),
                "events", events.stream()
                        .filter(event -> event instanceof Message)
                        .filter(not(SynchronousResponse.class::isInstance))
                        .map(event -> (Message) event)
                        .map(message -> String.format("[%s] -> %s", message.getFrom(), message.getTo()))
                        .distinct()
                        .collect(toList())
        ));
        return writer.toString();
    }
}
