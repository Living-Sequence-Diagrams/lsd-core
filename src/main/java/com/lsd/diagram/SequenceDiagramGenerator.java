package com.lsd.diagram;

import com.lsd.IdGenerator;
import com.lsd.events.SequenceEvent;
import com.lsd.properties.LsdProperties;
import com.lsd.report.model.Diagram;
import com.lsd.report.model.Participant;
import com.mitchellbosecke.pebble.PebbleEngine;
import com.mitchellbosecke.pebble.template.PebbleTemplate;
import lombok.Builder;
import lombok.SneakyThrows;

import java.io.StringWriter;
import java.io.Writer;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

import static com.lsd.properties.LsdProperties.DIAGRAM_THEME;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;
import static org.apache.commons.collections4.ListUtils.partition;

@Builder
public class SequenceDiagramGenerator {
    private final PebbleEngine engine = new PebbleEngine.Builder().autoEscaping(false).build();
    private final PebbleTemplate compiledTemplate = engine.getTemplate("templates/sequence-uml.peb");

    private final Set<String> includes;
    private final List<Participant> participants;
    private final List<SequenceEvent> events;
    private final IdGenerator idGenerator;

    public Optional<Diagram> diagram(int maxEventsPerDiagram) {
        if (events.isEmpty()) {
            return Optional.empty();
        }
        var uml = partition(events, maxEventsPerDiagram).stream()
                .map(this::generateSequenceUml)
                .collect(joining(System.lineSeparator()));

        String svg = generateSequenceSvg(uml);
        return Optional.of(Diagram.builder()
                .id(idGenerator.next())
                .uml(uml)
                .svg(svg)
                .build());
    }

    private String generateSequenceSvg(String markup) {
        SvgConverter svgConverter = new SvgConverter();

        return svgConverter.convert(markup);
    }

    @SneakyThrows
    private String generateSequenceUml(List<SequenceEvent> events) {
        Writer writer = new StringWriter();
        compiledTemplate.evaluate(writer, Map.of(
                "theme", LsdProperties.get(DIAGRAM_THEME),
                "includes", includes,
                "participants", participants.stream().distinct().collect(toList()),
                "events", events.stream()
                        .map(SequenceEvent::toMarkup)
                        .collect(toList())
        ));
        return writer.toString();
    }
}
