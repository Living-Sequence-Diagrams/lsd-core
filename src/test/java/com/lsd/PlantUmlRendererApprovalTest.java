package com.lsd;

import com.lsd.domain.scenario.diagram.DiagramGenerator;
import com.lsd.domain.scenario.events.SimpleInteraction;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Set;

class PlantUmlRendererApprovalTest {

    private DiagramGenerator diagramGenerator = DiagramGenerator.builder()
            .participants(List.of())
            .includes(Set.of())
            .events(List.of(
                    SimpleInteraction.builder()
                            .id("1")
                            .label("Sending a request")
                            .from("A")
                            .to("B")
                            .data("Please help?")
                            .build(),
                    SimpleInteraction.builder()
                            .id("2")
                            .label("Sending a response")
                            .from("B")
                            .to("A")
                            .data("OK")
                            .build()
            )).build();

    @Test
    void renderSequenceUml() {
        Approvals.verify(diagramGenerator.sequenceDiagram().getUml());
    }

    @Test
    void renderSequenceSvg() {
        Approvals.verifyXml(diagramGenerator.sequenceDiagram().toSvg());
    }
}