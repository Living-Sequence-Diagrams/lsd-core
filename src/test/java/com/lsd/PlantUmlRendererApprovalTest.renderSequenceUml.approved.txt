@startuml

hide unlinked

  !include <tupadr3/font-awesome-5/database>
  !include <tupadr3/font-awesome-5/clock>

  actor Arnold as "Arnie"
  boundary Nick
  control AControl
  collections SomeCollection
  entity Entity
  participant Party
  queue Party
  database Derek as "D for danger"

  A -> B:[[#1 Sending a request]]
  B -->> A:[[#2 Sending a response]]

@enduml